src/parser/control_parser.rs

```diff
@@ -41,110 +48,117 @@ fn parse_if_statement(lexer: &mut ParseSession) -> Result<Statement, Diagnostic>
     let mut else_block = Vec::new();
 
     if lexer.last_token == KeywordElse {
-        else_block.append(&mut parse_body_in_region(lexer, vec![KeywordEndIf])?);
+        else_block.append(&mut parse_body_in_region(lexer, vec![KeywordEndIf]));
     }
 
     let end = lexer.last_range.end;
 
-    Ok(Statement::IfStatement {
+    Statement::IfStatement {
         blocks: conditional_blocks,
         else_block,
         location: SourceRange::new(start..end),
-    })
+    }
 }
 
-fn parse_for_statement(lexer: &mut ParseSession) -> Result<Statement, Diagnostic> {
+fn parse_for_statement(lexer: &mut ParseSession) -> Statement {
     let start = lexer.range().start;
     lexer.advance(); // FOR
 
-    let counter_expression = parse_reference(lexer)?;
-    lexer.expect(KeywordAssignment)?; // :=
+    let counter_expression = parse_reference(lexer);
```

Comment:
If you formulate this as follows I think you can avoid the expect: 
After the FOR keyword, you start a region that ends with DO
Inside this region you will get tokens for the Assignment, TO and BY. I think you can treat these as consume or report instances. 
You would need to cover these by tests though (The for is an assignment, followed by a TO Reference, followed optionally by a BY reference and then closed by a DO)

Reply:
I think its a similar situation like in the first comment - but I'm not sure if I got everything right from your suggestion.
It would also be doable like the while again

```
let ref = match parse_reference(...) {
    Err(..) => report meaningful error in the current context, return empty-statement
    Ok(it) => it
};
allow(TO);
... 
```
maybe you could even create a fancy method or macro for it